src/main/scala/org/phenoscape/kb/Graph.scala

```diff
@@ -100,44 +102,68 @@ object Graph {
-    import scalaz._
-    import Scalaz._
+  def ancestorMatrix(terms: Set[IRI], relations: Set[IRI], termPathOpt: Option[Path]): Future[AncestorMatrix] =
     if (terms.isEmpty) Future.successful(AncestorMatrix(""))
     else {
-      val valuesElements = terms.map(t => sparql" $t ").reduce(_ + _)
+      val termsElements = terms.map(t => sparql" $t ").reduceOption(_ + _).getOrElse(sparql"")
+      val relationsElements = relations.map(r => sparql" $r ").reduceOption(_ + _).getOrElse(sparql"")
+      val queryPattern = termPathOpt match {
+        case Some(path) => sparql""" ?term $path ?class . ?class ?relation ?subsumer ."""
+        case None       => sparql""" ?term ?relation ?subsumer . """
+      }
+
       val query =
         sparql"""
-       SELECT DISTINCT ?term ?ancestor
+       SELECT DISTINCT ?term ?relation ?subsumer 
        FROM $KBClosureGraph
+       FROM $KBRedundantRelationGraph
        WHERE {
-         VALUES ?term { $valuesElements }
-         ?term $rdfsSubClassOf ?ancestor .
-         FILTER(?ancestor != $owlThing)
+         VALUES ?term { $termsElements }
+         VALUES ?relation { $relationsElements }
+         $queryPattern
+         FILTER(?subsumer != $owlThing)
        }
           """
-      val futurePairs = App.executeSPARQLQueryString(query.text,
-                                                     qs => {
-                                                       val term = qs.getResource("term").getURI
-                                                       val ancestor = qs.getResource("ancestor").getURI
-                                                       (term, ancestor)
-                                                     })
+
+      print(" \n ***" + query.text)
+
+      val futurePairs = App.executeSPARQLQueryString(
+        query.text,
+        qs => {
+          val term = qs.getResource("term").getURI
+          val relation = qs.getResource("relation").getURI
+          val subsumer = qs.getResource("subsumer").getURI
+          (term, relation, subsumer)
+        }
+      )
       for {
         pairs <- futurePairs
       } yield {
         val termsSequence = terms.map(_.toString).toSeq.sorted
         val header = s",${termsSequence.mkString(",")}"
-        val groupedByAncestor = pairs.groupBy(_._2)
-        val valuesLines = groupedByAncestor.map {
-          case (ancestor, ancPairs) =>
-            val termsForAncestor = ancPairs.map(_._1).toSet
-            val values = termsSequence.map(t => if (termsForAncestor(t)) "1" else "0")
-            s"$ancestor,${values.mkString(",")}"
+        val termSubsumerPairs = pairs.map(p => (p._1 -> (p._2, p._3)))
+
+        // creates Map(term -> virtualIRI(relation, subsumer))
+        val baseIRI = "http://purl.org/phenoscape/term/virtual/owl/ObjectSomeValuesFrom"
+        val termToRelSubsumerSeq = termSubsumerPairs.map { case (term, pairs) =>
+          val virtualTermIRI = RelationalTerm(IRI.create(pairs._1), IRI.create(pairs._2)).iri
```

Comment:
Don't create a virtual term for subClassOf.

Reply:
New test results after update: https://github.com/phenoscape/phenoscape-kb-services/issues/393#issuecomment-825078263